routers/private/hook.go

```diff
@@ -98,7 +98,7 @@ func HookPreReceive(ctx *macaron.Context) {
 
 		canPush := false
 		if isDeployKey {
-			canPush = protectBranch.WhitelistDeployKeys
+			canPush = protectBranch.CanPush && (!protectBranch.EnableWhitelist || protectBranch.WhitelistDeployKeys)
```

Comment:
Previously the fact that people were listed in the whitelist wasn't an impediment to push using the deploy key. If I get this correctly, now deploy keys can't be used if there is a whitelist?

Reply:
My intention is that deploy keys can push if:
- Form is set to _Enable Push_. This corresponds to CanPush=True and EnableWhitelist=False
- Form is set to _Whitelist Restricted Push_ and _Whitelist deploy keys..._ is ticked. This corresponds to CanPush=True, EnableWhitelist=True and WhitelistDeployKeys=True

Reply:
I don't use deploy keys so I'm not familiar with its use case, but this will be a breaking change them, I believe?

Reply:
Neither do I use deploy keys, but I don't think the behavior will change with this PR. I have added the term `CanPush` as requirement for any pushing (migration step handles that). `protectBranch.WhitelistDeployKeys` is the existing condition for deploy key pushing. The new condition is that deploy keys can still push if whitelist is not used (option **Enable Push**), i.e. `!protectBranch.EnableWhitelist`.

Reply:
Potentially if `protectBranch.EnableWhitelist` is false and `protectedBranch.WhitelistDeployKeys` is true it would be a change of behavior. But that would only be possible because we don't clear existing flags and I don't think that is correct behavior.

Reply:
Sorry, I see now what you mean. 👍